include/material_models/J2Plasticity.h

```diff
@@ -0,0 +1,278 @@
+#ifndef J2PLASTICITY_H
+#define J2PLASTICITY_H
+
+#include "matmodel.h"
+#include "solver.h"
+
+class J2Plasticity : public MechModel {
+  public:
+    J2Plasticity(vector<double> l_e, map<string, vector<double>> materialProperties)
+        : MechModel(l_e)
+    {
+        try {
+            bulk_modulus  = materialProperties["bulk_modulus"];
+            shear_modulus = materialProperties["shear_modulus"];
+            yield_stress  = materialProperties["yield_stress"];                  // Initial yield stress
+            K             = materialProperties["isotropic_hardening_parameter"]; // Isotropic hardening parameter
+            H             = materialProperties["kinematic_hardening_parameter"]; // Kinematic hardening parameter
+            eta           = materialProperties["viscosity"];                     // Viscosity parameter
+            dt            = materialProperties["time_step"][0];                  // Time step
+        } catch (const std::out_of_range &e) {
+            throw std::runtime_error("Missing material properties for the requested material model.");
+        }
+        n_mat = bulk_modulus.size();
+
+        Matrix<double, 6, 6> *Ce      = new Matrix<double, 6, 6>[n_mat];
+        Matrix<double, 6, 6>  topLeft = Matrix<double, 6, 6>::Zero();
+        topLeft.topLeftCorner(3, 3).setConstant(1);
+
+        kapparef_mat = Matrix<double, n_str, n_str>::Zero();
+        for (int i = 0; i < n_mat; ++i) {
+            Ce[i] = 3 * bulk_modulus[i] * topLeft +
+                    2 * shear_modulus[i] * (-1.0 / 3.0 * topLeft + Matrix<double, 6, 6>::Identity());
+            kapparef_mat += Ce[i];
+        }
+        kapparef_mat /= n_mat;
+
+        // Allocate the member matrices/vectors for performance optimization
+        sqrt_two_over_three = sqrt(2.0 / 3.0);
+        sigma_trial_n1.setZero();
+        dev.setZero();
+        qbar_trial_n1.setZero();
+        n.setZero();
+        eps_elastic.setZero();
+        dev_minus_qbar.setZero();
+    }
+
+    virtual void initializeInternalVariables(ptrdiff_t num_elements, int num_gauss_points) override
+    {
+        // Initialize plastic strain and other internal variables
```

Comment:
Maybe add documentation for new users here what the different variants "t" and "n1" mean

Reply:
Good suggestion. Will do it!